wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala

```diff
@@ -0,0 +1,283 @@
+package org.bitcoins.wallet.internal
+
+import org.bitcoins.wallet.LockedWallet
+import org.bitcoins.core.protocol.transaction.Transaction
+import org.bitcoins.wallet.models._
+import scala.concurrent.Future
+import org.bitcoins.core.protocol.transaction.TransactionOutput
+import org.bitcoins.wallet.api.AddUtxoSuccess
+import org.bitcoins.wallet.api.AddUtxoError
+import org.bitcoins.core.number.UInt32
+import org.bitcoins.core.util.FutureUtil
+
+/** Provides functionality for processing transactions. This
+  * includes importing UTXOs spent to our wallet, updating
+  * confirmation counts and marking UTXOs as spent when
+  * spending from our wallet
+  */
+private[wallet] trait TransactionProcessing { self: LockedWallet =>
+  /////////////////////
+  // Public facing API
+
+  /** @inheritdoc */
+  override def processTransaction(
+      transaction: Transaction,
+      confirmations: Int): Future[LockedWallet] = {
+    logger.info(
+      s"Processing transaction=${transaction.txIdBE} with confirmations=$confirmations")
+    processTransactionImpl(transaction, confirmations).map {
+      case ProcessTxResult(outgoing, incoming) =>
+        logger.info(
+          s"Finished processing of transaction=${transaction.txIdBE}. Relevant incomingTXOs=${incoming.length}, outgoingTXOs=${outgoing.length}")
+        this
+    }
+
+  }
+
+  private[wallet] case class ProcessTxResult(
+      updatedIncoming: List[IncomingWalletTXO],
+      updatedOutgoing: List[IncomingWalletTXO])
+
+  /////////////////////
+  // Internal wallet API
+
+  /**
+    * Processes TXs originating from our wallet.
+    * This is called right after we've signed a TX,
+    * updating our UTXO state.
+    */
+  private[wallet] def processOurTransaction(
+      transaction: Transaction,
+      confirmations: Int) = {
+    logger.info(
+      s"Processing TX from our wallet, transaction=${transaction.txIdBE} with confirmations=$confirmations")
+    processTransactionImpl(transaction, confirmations).map { result =>
+      val txid = transaction.txIdBE
+      val changeOutputs = result.updatedIncoming.length
+      val spentOutputs = result.updatedOutgoing.length
+
+      logger.info(
+        s"Processing of internal transaction=$txid resulted in changeOutputs=$changeOutputs and spentUTXOs=$spentOutputs")
+      result
+    }
+  }
+
+  /////////////////////
+  // Private methods
+
+  /** Does the grunt work of processing a TX.
+    * This is called by either the internal or public TX
+    * processing method, which logs and transforms the
+    * output fittingly.
+    */
+  private def processTransactionImpl(
+      transaction: Transaction,
+      confirmations: Int): Future[ProcessTxResult] = {
+
+    val incomingTxoFut: Future[Vector[IncomingWalletTXO]] =
+      incomingTxoDAO
+        .findTx(transaction)
+        .flatMap {
+          // no existing elements found
+          case Vector() =>
+            processNewIncomingTx(transaction, confirmations)
```

Comment:
I don't think I really like the distinction between `processNewIncomingTx` and `processExistingIncomingTxo`, I think the behavior seems redundant and it wouldn't be hard to consolidate this down to one method. Also why does `processNewIncomingTx` take a confirmation parameter? Is this because we aren't making strong guarantees about when the wallet is receiving blocks? I.e. the wallet is offline for 2 days before we sync it back up?

Reply:
The reason for keeping them as two different method is just to keep the method smaller. Can change it if you feel strongly about it.

The reason for having a confirmations param in `processNewIncomingTx` is exactly what you posted yourself.